execution/executor/src/components/in_memory_state_calculator_v2.rs

```diff
@@ -0,0 +1,253 @@
+use aptos_storage_interface::{cached_state_view::StateCache, state_delta::StateDelta};
+use aptos_types::{
+    account_address::AccountAddress,
+    account_config::CORE_CODE_ADDRESS,
+    account_view::AccountView,
+    epoch_state::EpochState,
+    state_store::{
+        state_key::StateKey, state_storage_usage::StateStorageUsage, state_value::StateValue,
+    },
+    transaction::{Transaction, Version},
+    write_set::TransactionWrite,
+};
+use arr_macro::arr;
+use dashmap::DashMap;
+use rayon::prelude::*;
+use std::collections::HashMap;
+
+struct CoreAccountStateView<'a> {
+    base: &'a [DashMap<StateKey, Option<StateValue>>; 16],
+    updates: &'a HashMap<&'a StateKey, &'a Option<StateValue>>,
+}
+
+impl<'a> CoreAccountStateView<'a> {
+    pub fn new(
+        base: &'a [DashMap<StateKey, Option<StateValue>>; 16],
+        updates: &'a HashMap<&'a StateKey, &'a Option<StateValue>>,
+    ) -> Self {
+        Self { base, updates }
+    }
+}
+
+impl<'a> AccountView for CoreAccountStateView<'a> {
+    fn get_state_value(&self, state_key: &StateKey) -> Result<Option<Vec<u8>>> {
+        if let Some(v_opt) = self.updates.get(state_key) {
+            return Ok(v_opt.as_ref().map(|x| x.bytes().to_vec()));
+        }
+        if let Some(v_opt) = self.base[state_key.get_shard_id() as usize]
+            .get(state_key)
+            .as_ref()
+        {
+            return Ok(v_opt.as_ref().map(|x| x.bytes().to_vec()));
+        }
+        Ok(None)
+    }
+
+    fn get_account_address(&self) -> Result<Option<AccountAddress>> {
+        Ok(Some(CORE_CODE_ADDRESS))
+    }
+}
+
+/// Helper class for calculating state changes after a block of transactions are executed.
+pub struct InMemoryStateCalculatorV2 {
+    ///// These don't change during the calculation.
+    // This makes sure all in-mem nodes seen while proofs were fetched stays in mem during the
+    // calculation
+    _frozen_base: FrozenSparseMerkleTree<StateValue>,
+    proof_reader: ProofReader,
+
+    // The states at base version.
+    sharded_state_cache: [DashMap<StateKey, Option<StateValue>>; 16],
+    next_version: Version,
+    usage: StateStorageUsage,
+
+    //// These changes whenever make_checkpoint() happens.
+    checkpoint: SparseMerkleTree<StateValue>,
+    checkpoint_version: Option<Version>,
+    // This doesn't need to be frozen since `_frozen_base` holds a ref to the oldest ancestor
+    // already, but frozen SMT is used here anyway to avoid exposing the `batch_update()` interface
+    // on the non-frozen SMT.
+    latest: FrozenSparseMerkleTree<StateValue>,
+}
+
+impl InMemoryStateCalculatorV2 {
+    pub fn new(base: &StateDelta, state_cache: StateCache) -> Self {
+        let StateCache {
+            frozen_base,
+            state_cache,
+            proofs,
+        } = state_cache;
+        // TODO(grao): Only a base tree and version are needed because the current should be the
+        // same as base.
```

Comment:
I guess I've lost the context.. so why?

Reply:
Because the assert! at line 97 and 98 will always hold, or am I missing anything?